Clean up qsv filters

xhaihao opened the pull request:
Clean-up scale_qsv and deinterlace_qsv in this patchset. In addition, I added a few features to vpp_qsv / scale_qsv / deinterlace_qsv.

xuguangxin commented on libavfilter/vf_scale_qsv.c:
```diff
@@ -534,9 +208,43 @@ static int qsvscale_config_props(AVFilterLink *outlink)
+    if (inlink->w != outlink->w ||
+        inlink->h != outlink->h ||
+        in_format != s->format) {
+        ret = ff_qsvvpp_init(ctx, &param);
```
Not related to this. But, we may need to add some passthrough skip mode. 
for example, the customer may need to add subtitles on decode output.

xhaihao replied:
I'm sorry I didn't catch your point. User may not use this filter if they want 'passthrough skip mode'

xuguangxin replied:
How to draw something on decode output? We need some filter copy input to output.

xhaihao replied:
I think you should use overlay_qsv instead of scaling_qsv for your case.

xuguangxin commented on libavfilter/vf_scale_qsv.c:
```diff
@@ -558,67 +266,20 @@ static int qsvscale_config_props(AVFilterLink *outlink)
+    } else {
+        if (in->pts != AV_NOPTS_VALUE)
+            in->pts = av_rescale_q(in->pts, link->time_base, outlink->time_base);
+        ret = ff_filter_frame(outlink, in);
```
why a scale filter needs pts conversion?

xhaihao replied:
You are right, pts conversion is not needed here.  The intention of adding the code here is to make it similar to the function in vf_vpp_qsv.c, and then re-using all functions in vf_vpp_qsv.c for scale_qsv filter.

xuguangxin commented on libavfilter/vf_scale_qsv.c:
```diff
@@ -558,67 +266,20 @@ static int qsvscale_config_props(AVFilterLink *outlink)
-    if (err < 0) {
-        ret = ff_qsvvpp_print_error(ctx, err, "Error synchronizing the operation");
-        goto fail;
+    if (qsv->session) {
```
Could we have dedicate flag for pass-through?
or we can add some comments?

xhaihao replied:
Sure I will add some comment.

xuguangxin commented on libavfilter/qsvvpp.c:
```diff
@@ -837,8 +841,17 @@ int ff_qsvvpp_filter_frame(QSVVPPContext *s, AVFilterLink *inlink, AVFrame *picr
+        if (ret == MFX_ERR_MORE_SURFACE &&
+            out_frame->surface.Data.TimeStamp == MFX_TIMESTAMP_UNKNOWN &&
+            s->last_in_pts != AV_NOPTS_VALUE)
+            dpts = (in_frame->frame->pts - s->last_in_pts) / 2;
```
In frame conversion case, we may need 1 frame in N frame out, 
/ 2 may not right

xhaihao replied:
I didn't try FRC. It is possible the timestamp returned from the SDK is valid, otherwise FRC should be wrong even if without this patch. Anyway I will try FRC after cleanup the code.

xuguangxin commented on libavfilter/vf_deinterlace_qsv.c:
```diff
@@ -115,441 +70,54 @@ static int qsvdeint_query_formats(AVFilterContext *ctx)
-    QSVFrame *frame, **last;
-
-    clear_unused_frames(s);
+    if (inlink->format == AV_PIX_FMT_QSV) {
```
could we have a function for this logical?

xhaihao replied:
This is not the final version. I will re-use all functions in vf_vpp_qsv.c for deinterlace_qsv filter too.

xhaihao commented:
@xuguangxin This PR is not done. Currently I use this PR to check whether it causes any regression, not for code review. I will add you and others to review this PR once it is done. I should may some key words (such as WIP) in the title.

xhaihao committed:
lavf/qsv: use QSVVPPContext as base context in vf_vpp_qsv/vf_overlay_qsv

The same members between QSVVPPContext and VPPContext are removed from
VPPContext, and async_depth is moved from QSVVPPParam to QSVVPPContext
so that all QSV filters using QSVVPPContext may support async depth. In
addition we may use QSVVPPContext as base context in other QSV filters
in the future.

xhaihao committed:
lavf/scale_qsv: simplify scale_qsv filter

Use QSVVPPContext as a base context of QSVScaleContext, hence we may
re-use functions defined for QSVVPPContext to manage MFX session for
scale_qsv filter too. Because system memory is taken into account in
QSVVVPPContext, we may add support for non-QSV pixel formats in the
future

xhaihao committed:
lavf/scale_qsv: don't need variables for constants in FFmpeg

PI, PHI and E are defined in FFmpeg

xhaihao committed:
lavf/vpp_qsv: add "a", "dar" and "sar" variables

Also fix the coding style for VAR index. This is in preparation for
re-using VPPContext for scale_qsv filter

xhaihao committed:
lavf/vpp_qsv: handle NULL pointer when evaluating an expression

This is in preparation for re-using VPPContext but with a different
option array for scale_qsv filter

xhaihao committed:
lavf/vpp_qsv: allow special values for the output dimensions

Special values are:
0 = original width/height
-1 = keep original aspect

This is in preparation for re-using VPPContext for scale_qsv filter

xhaihao committed:
lavf/vpp_qsv: factorize extra MFX configuration

This is in preparation for re-using VPPContext for scale_qsv filter

xhaihao committed:
lavf/vpp_qsv: pass scaling mode to the SDK

After this patch, the scaling mode will be passed to the SDK when the
scaling mode is not equal to the default mode. This is in preparation
for re-using VPPContext for scale_qsv filter

xhaihao committed:
lavf/vpp_qsv: add vpp_preinit callback

Set the expected default value for options in this callback, hence we
have the right values even if these options are not included in the
option arrray. This is in preparation for re-using VPPContext but with a
different option array for other QSV filters

xhaihao committed:
lavf/scale_qsv: re-use VPPContext for scale_qsv filter

All features are implemented in vpp_qsv filter, scale_qsv can be taken
as a special case of vpp_qsv filter now, we re-use VPPContext with a
different option arrary and pixel formats

xhaihao committed:
lavf/vpp_qsv: factor common QSV filter definition

xhaihao committed:
lavf/scale_qsv: add new options for scale_qsv filter

Allow user to set crop area and async depth

xhaihao committed:
lavf/scale_qsv: add more input / output pixel formats

NV12 and P010 are added

$ ffmpeg -init_hw_device qsv -c:v h264_qsv -i input.h264 -vf
"scale_qsv=format=p010" -f null -

xhaihao committed:
lavf/vpp_qsv: double the framerate for deinterlacing

xhaihao committed:
lavf/qsvvpp: avoid overriding the returned value

Currently the returned value from MFXVideoVPP_RunFrameVPPAsync() is
overridden, so the check of 'ret == MFX_ERR_MORE_SURFACE' is always
false when MFX_ERR_MORE_SURFACE is returned from
MFXVideoVPP_RunFrameVPPAsync()

xhaihao committed:
lavf/qsvvpp: set PTS for output frame

When the SDK returns MFX_ERR_MORE_SURFACE, the PTS is not set for the
output frame. We assign a PTS calculated from the input frame to the
output frame. After applying this patch, we may avoid the error below:

[null @ 0x56395cab4ae0] Application provided invalid, non monotonically
increasing dts to muxer in stream 0: 456 >= 0

Note this patch only fixes PTS issue when deinterlacing is enabled

xhaihao committed:
lavf/vpp_qsv: check output format string against NULL pointer

This is in preparation for re-using VPPContext but with a different
option array for deinterlacing_qsv filter

xhaihao committed:
lavf/deinterlace_qsv: simplify deinterlace_qsv filter

Like what we did for scale_qsv filter, we use QSVVPPContext as a base
context to manage MFX session for deinterlace_qsv filter

xhaihao committed:
lavf/deinterlace_qsv: re-use VPPContext for deinterlace_qsv filter

All features are implemented in vpp_qsv filter now, so deinterlace_qsv
can be taken as a specical case of vpp_qsv filter, we re-use VPPContext
with a different option array and pix formats for deinterlace_qsv filter

xhaihao committed:
lavf/deinterlace_qsv: add async_depth option

Allow user to set async depth for deinterlace_qsv

xhaihao committed:
lavf/deinterlace_qsv: add more input / output pixel formats

NV12 is added in system memory and the command below may work now.

$ ffmpeg -init_hw_device qsv -c:v h264_qsv -i input.h264 -vf
deinterlace_qsv -f null -

xhaihao committed:
lavf/vpp_qsv: allow user to set scaling mode for vpp_qsv filter

option 'scaling' accepts one of low_power and hq

$ ffmpeg -init_hw_device qsv -hwaccel qsv -c:v h264_qsv -i input.h264
-vf "vpp_qsv=scaling=hq" -f null -

xuguangxin commented on libavfilter/vf_vpp_qsv.c:
```diff
@@ -654,6 +654,16 @@ static const AVOption vpp_options[] = {
+    { "low_power", "low power mode",        0,             AV_OPT_TYPE_CONST,  { .i64 = MFX_SCALING_MODE_LOWPOWER}, INT_MIN, INT_MAX, FLAGS, "mode"},
+    { "hq",        "high quality mode",     0,             AV_OPT_TYPE_CONST,  { .i64 = MFX_SCALING_MODE_QUALITY},  INT_MIN, INT_MAX, FLAGS, "mode"},
+#else
+    { "scaling",      "(not supported)",     OFFSET(scaling_mode),    AV_OPT_TYPE_INT,    { .i64 = 0}, 0, INT_MAX, FLAGS, "mode"},
```
why we need the #else?

xhaihao replied:
User may use the same commandline w/wo  scaling mode support in MSDK

xuguangxin replied:
how about we use 0, 1, 2 and some use an array to map index to msdk value.
```
int scale_mode_map = {MFX_SCALING_MODE_DEFAULT, MFX_SCALING_MODE_LOWPOWER, MFX_SCALING_MODE_QUALITY];
vpp->scaling_mode = scaling_mode_map[scaling_mode]
```

xhaihao replied:
MFX_SCALING_MODE_DEFAULT, MFX_SCALING_MODE_LOWPOWER and MFX_SCALING_MODE_QUALITY have already been used in scale_qsv filter, so we should use them in vpp_qsv too. 

BTW MFX_SCALING_MODE_DEFAULT, MFX_SCALING_MODE_LOWPOWER and MFX_SCALING_MODE_QUALITY are 0, 1, 2 actually and user may use string (low_power, hq etc) to set the scaling mode.

xuguangxin commented on libavfilter/vf_vpp_qsv.c:
```diff
@@ -272,11 +270,38 @@ static int config_input(AVFilterLink *inlink)
+    if (oh == -1)
+        oh = av_rescale(ow, inlink->h, inlink->w);
+
+    if (ow > INT_MAX || oh > INT_MAX ||
```
ow is int, it will never > INT_MAX

xhaihao replied:
ow, oh should be int64_t, I will change it.

xhaihao replied:
Changed in the new version, thx.

xuguangxin commented on libavfilter/vf_vpp_qsv.c:
```diff
@@ -272,11 +270,38 @@ static int config_input(AVFilterLink *inlink)
+        oh = av_rescale(ow, inlink->h, inlink->w);
+
+    if (ow > INT_MAX || oh > INT_MAX ||
+        (oh * inlink->w) > INT_MAX  ||
```
could this check by the driver only?

xhaihao replied:
scale_qsv has this check, so I added this check in vpp_qsv too. I'm not sure whether the driver has this check.

xuguangxin commented on libavfilter/vf_vpp_qsv.c:
```diff
@@ -460,11 +486,27 @@ static int config_output(AVFilterLink *outlink)
 
+    if (vpp->scaling_mode) {
+#ifdef QSV_HAVE_SCALING
+        memset(&vpp->scaling_conf, 0, sizeof(mfxExtVPPScaling));
```
How about we define some macro for this?
```
#define INIT_VPP_PARAM(param, id) \
memset(param, 0, sizeof(*param)); \
param->Header...\
param.ext_buf[param.num_ext_buf++] = (mfxExtBuffer*)param;
```

xhaihao replied:
I may re-factor the code out, thanks for the comment.

xhaihao replied:
changed in the new version, thx.

xuguangxin commented on libavfilter/vf_vpp_qsv.c:
```diff
@@ -559,48 +693,65 @@ static int query_formats(AVFilterContext *ctx)
+    { "low_power", "low power mode",        0,             AV_OPT_TYPE_CONST,  { .i64 = MFX_SCALING_MODE_LOWPOWER}, INT_MIN, INT_MAX, FLAGS, "mode"},
+    { "hq",        "high quality mode",     0,             AV_OPT_TYPE_CONST,  { .i64 = MFX_SCALING_MODE_QUALITY},  INT_MIN, INT_MAX, FLAGS, "mode"},
+#else
+    { "mode",      "(not supported)",     OFFSET(scaling_mode),    AV_OPT_TYPE_INT,    { .i64 = 0}, 0, INT_MAX, FLAGS, "mode"},
```
same comment as above

xhaihao replied:
See the reply for scaling.

xuguangxin commented on libavfilter/vf_vpp_qsv.c:
```diff
@@ -559,48 +693,65 @@ static int query_formats(AVFilterContext *ctx)
+static const AVOption qsvdeint_options[] = {
+    { "mode", "set deinterlace mode", OFFSET(deinterlace),   AV_OPT_TYPE_INT, {.i64 = MFX_DEINTERLACING_ADVANCED}, MFX_DEINTERLACING_BOB, MFX_DEINTERLACING_ADVANCED, FLAGS, "mode"},
+    { "bob",   "bob algorithm",                  0, AV_OPT_TYPE_CONST,      {.i64 = MFX_DEINTERLACING_BOB}, MFX_DEINTERLACING_BOB, MFX_DEINTERLACING_ADVANCED, FLAGS, "mode"},
+    { "advanced", "Motion adaptive algorithm",   0, AV_OPT_TYPE_CONST, {.i64 = MFX_DEINTERLACING_ADVANCED}, MFX_DEINTERLACING_BOB, MFX_DEINTERLACING_ADVANCED, FLAGS, "mode"},
```
advanced maybe too vague， how about define it as motion_adaptive

xhaihao replied:
advanced has been used in deinterlace_qsv. I prefer to use 'advanced', otherwise it is easy to mislead user that there is difference between 'advanced' and 'motion_adaptive'.

xuguangxin commented on libavfilter/qsvvpp.c:
```diff
@@ -874,5 +885,7 @@ int ff_qsvvpp_filter_frame(QSVVPPContext *s, AVFilterLink *inlink, AVFrame *picr
         }
     } while(ret == MFX_ERR_MORE_SURFACE);
 
+    s->last_in_pts = in_frame->frame->pts;
```
how to handle new segments

xhaihao replied:
Sorry I didn't catch your question, could you please describe your question in detail?

xuguangxin replied:
for some clips, the timestamp may not continue. not sure how to handle this in filter.

xhaihao replied:
Did you mean the input frame doesn't have a valid pts? We don't handle this case because the output and input frames have the same pts setting except deinterlacing.

xuguangxin commented on libavfilter/vf_vpp_qsv.c:
```diff
@@ -460,11 +486,27 @@ static int config_output(AVFilterLink *outlink)
+#else
+        av_log(ctx, AV_LOG_WARNING, "The scaling_mode option is "
+            "not supported with this MSDK version.\n");
+        vpp->scaling_mode = MFX_SCALING_MODE_DEFAULT;
```
is MFX_SCALING_MODE_DEFAULT defined in #else branch?

xhaihao replied:
Thanks for catching this. The doc didn't mention when MFX_SCALING_MODE_DEFAULT was added. But according to https://github.com/Intel-Media-SDK/MediaSDK/blob/master/doc/mediasdk-man.md, MFX_EXTBUFF_VPP_SCALING was added in the SDK API 1.19. I guess MFX_SCALING_MODE_DEFAULT was added in 1.19 too.

xhaihao replied:
Updated  in the new version, thanks.

xuguangxin commented:
not sure why fate failed.

xuguangxin commented:
@uartie , we never merge commit for this repo. not sure we can turn off the rebase and merge button for everyone or not?
thanks

xhaihao commented:
@uartie the check failed because flv-demux failed in fate.   The check passed before and this patch changed qsv filters only which is not used in fate

BTW `make fate-flv-demux V=2 Keep=1` failed with upstream FFmpeg, so it should be a regression in FFmpeg.

xhaihao commented:
It passed with commit 7539a01bb0ea24998e1aa87706edf9785ddd3ac6.

xhaihao commented:
The regression was caused by 958bea5248f87116b0dd080461aa70c14ea86cf0.  Was Enigma_Principles_of_Lust-part.flv updated in the fate server?

xhaihao commented:
I am pinging the author for this issue, see http://ffmpeg.org/pipermail/ffmpeg-devel/2021-May/280113.html

uartie commented:
> @uartie , we never merge commit for this repo. not sure we can turn off the rebase and merge button for everyone or not?
> thanks

GitHub requires at least one option to be enabled :(.  We just have to avoid using it.

uartie commented:
> The regression was caused by 958bea5. Was Enigma_Principles_of_Lust-part.flv updated in the fate server?

Our fate asset mirror runs every 15 minutes and test agents sync every 30 minutes.  But case is still failing... so must be upstream issue.

> I am pinging the author for this issue, see http://ffmpeg.org/pipermail/ffmpeg-devel/2021-May/280113.html

Thanks for pinging... hope it can be fixed soon.  Can you file a trac.ffmpeg.org bug if they don't respond in a day or two?

xhaihao commented:
> Thanks for pinging... hope it can be fixed soon. Can you file a trac.ffmpeg.org bug if they don't respond in a day or two?

@uartie @xuguangxin Did you receive my email from the mailing list? I didn't this email myself :(

uartie commented:
> > Thanks for pinging... hope it can be fixed soon. Can you file a trac.ffmpeg.org bug if they don't respond in a day or two?
> 
> @uartie @xuguangxin Did you receive my email from the mailing list? I didn't this email myself :(

No, I did not receive the email either :(

xhaihao commented:
I filed http://trac.ffmpeg.org/ticket/9235 to track this issue.

Semmer2 commented:
> I filed http://trac.ffmpeg.org/ticket/9235 to track this issue.

hi, Haihao,
just fyi
I came across the same issue with CPU i7-8700k (CFL), but @dingintel2021 Wenlong passed the fate test with CPU i5-8500 (CFL)

xhaihao commented:
@Semmer2 Thanks for the info, do you have the same software stack?

uartie commented:
@xhaihao @Semmer2 http://trac.ffmpeg.org/ticket/9235 is fixed.  Please rebase your PRs.

xhaihao commented:
> @xhaihao @Semmer2 http://trac.ffmpeg.org/ticket/9235 is fixed. Please rebase your PRs.

@uartie, thanks for reminding me, I rebased the patchset.

xhaihao commented:
the patchset was sent to FFmpeg mailing list for review http://ffmpeg.org/pipermail/ffmpeg-devel/2021-May/280252.html
